### PR TITLE
Add the Asset Store component

### PR DESCRIPTION
**Pull request by kazydek**
**Description**

Changes proposed in this pull request:

- Added documentation on a new Asset Store component
- Prepared configuration for the new component in docs

**Related issue(s)**
See also #2379

### CODE REVIEW

File: docs/asset-store/docs/06-01-asset.md [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+
+| Parameter   |      Mandatory      |  Description |
+|:----------:|:-------------:|:------|
+| **metadata.name** |    **YES**   | Specifies the name of the CR. |
+| **metadata.namespace** |    **YES**   | Defines the Namespace in which the CR is available. |
+| **spec.source.mode** |    **YES**   | Specifies if the asset is a single file, zip or tar, or a `index.yaml` file. For details on the **index** mode, see [this](#sample-index-file) section. |
+| **spec.source.url** |    **YES**   | Specifies the location of the file. |
+| **bucketref.name** |    **YES**   | Specifies the name of the bucket for storing the asset. |
+| **spec.source.validationwebhookservice** |    **NO**   | Provides specification of the validation webhook service. |
```
Comment (derberg) [inline docs/asset-store/docs/06-01-asset.md]: we are missing description of `spec.source.validationwebhookservice.name` and others like namespace, endpoint, metadata

File: docs/asset-store/docs/06-01-asset.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+| **spec.source.mutationwebhookservice** |    **NO**   | Provides specification of the validation webhook service. |
+
```
Comment (derberg) [inline docs/asset-store/docs/06-01-asset.md]: we also need detailed description of status fields here

File: docs/asset-store/docs/06-02-bucket.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+## Custom resource parameters
+
+This table lists all the possible parameters of a given resource together with their descriptions:
+
+
+| Parameter   |      Mandatory      |  Description |
+|:----------:|:-------------:|:------|
+| **metadata.name** |    **YES**   | Specifies the name of the CR. |
+| **metadata.namespace** |    **NO**   | Specifies the Namespace in which the CR is available. |
```
Comment (derberg) [inline docs/asset-store/docs/06-02-bucket.md]: please talk to @pkosiec there are already 2 spec fields implemented, not like we assumed at the beginning, spec.policy and spec.region

File: docs/asset-store/docs/01-01-asset-store.md [hunk 4]
```diff
@@ -0,0 +1,12 @@
+---
+title: Asset Store
+type: Overview
+---
+
+Asset Store is a generic, Kubernetes-native solution for storing documentation, images, API specifications, and client-side applications.
```
[the remark this answers is not in the document]
Comment (pkosiec) [reply docs/asset-store/docs/01-01-asset-store.md]: I would suggest `...for storing assets, such as documentation, images...`

File: docs/asset-store/docs/02-01-asset-store.md [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+- **Asset custom resource** is an obligatory [custom resource](./06-01-asset.md) (CR) in which you define the asset you want to store in a given storage bucket. Its definition requires the asset name and mode, the name of the Namespace in which it is available, the address to its Internet location, and the name of the bucket in which you want to store it. Optionally, you can specify the validation and mutation operations that the asset must undergo before it is stored.
+
+- **Asset Controller** validates if the bucket specified in the Asset CR exists, fetches the asset based on the information provided in the Asset CR, unpacks it if necessary, and performs optional asset validation and mutation. It finally uploads the asset into the specified Minio (Gateway) bucket.
+
+- **Bucket custom resource** is an obligatory [custom resource](./06-02-bucket.md) in which you define the name of the bucket for storing assets.
+
+- **Bucket Controller** listens to the events from the Bucket CR and checks if any new bucket was created or an existing one was deleted. It later uploads the new bucket onto Minio (Gateway) or deletes the bucket from it.
```
Comment (pkosiec) [inline docs/asset-store/docs/02-01-asset-store.md]: `It later uploads the new bucket onto Minio` -> `It later creates the new  Minio (Gateway) bucket`? Because in the `Asset Controller` section we have: `It finally uploads the asset into the specified Minio (Gateway) bucket`, so this change would make sense to me.

File: docs/asset-store/docs/03-02-bucket-cr-lifecycle.md [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+title: Bucket Custom Resource Lifecycle
+type: Details
+---
+
+Read about the lifecycle of the Bucket Custom Resource (CR) and how other resources react to its creation and removal.
+
+
+## Create a Bucket CR
+
+When you create a Bucket CR, the Bucket Controller receives and event on the new CR and creates a bucket with the **name** specified in the CR is created in the Minio Gateway storage under the `ns-{NAMESPACE_NAME}-{CR_NAME}` location. The status of the CR contains a reference URL to the created bucket.
```
Comment (pkosiec) [inline docs/asset-store/docs/03-02-bucket-cr-lifecycle.md]:
```suggestion
When you create a Bucket CR, the Bucket Controller receives the event and creates a bucket with the **name** specified in the CR. It is created in the Minio Gateway storage under the `ns-{NAMESPACE_NAME}-{CR_NAME}` location. The status of the CR contains a reference URL to the created bucket.
```

`the event` - I'm not sure. But this is all about the "bucket CR has been created" event

File: docs/asset-store/docs/06-02-bucket.md (hunk 3 above)
Comment (pkosiec) [reply docs/asset-store/docs/06-02-bucket.md]: Yes, there are two additional fields in `spec`:

1. `location` - location of the bucket, where it should be created. Here are listed all valid locations: https://docs.minio.io/docs/golang-client-api-reference.html#MakeBucket
If the field is empty, it means the default location will be used. So yeah, this is optional field.

Maybe I could make an enum with these values...

2. `policy` - here the things are more complicated. This is optional field as well. Default policy is empty, which means there is no policy for the bucket. If an user wants to change the policy, (s)he needs to provide policy in string:

something like that:

`{"Version": "2012-10-17","Statement": [{"Action": ["s3:GetObject"],"Effect": "Allow","Principal": {"AWS": ["*"]},"Resource": ["arn:aws:s3:::my-bucketname/*"],"Sid": ""}]}`

To be honest I need to dig into this topic, because I'm not able to provide more pieces of information about the whole `policy` topic. Things had been much simpler a year ago - the API has been changed recently.

File: docs/asset-store/docs/06-02-bucket.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+## Sample custom resource
+
+This is a sample resource that defines the storage bucket configuration.
+
+```
+apiVersion: assetstore.kyma-project.io/v1alpha1
+kind: Bucket
+metadata:
+  name: my-bucket
+  namespace: default
```
Comment (pkosiec) [inline docs/asset-store/docs/06-02-bucket.md]:
```
apiVersion: assetstore.kyma-project.io/v1alpha1
kind: Bucket
metadata:
  name: my-bucket
  namespace: default
spec:
  location: "us-east-1"
  policy: `
     {
      "Version":"2012-10-17",
      "Statement":[
         {
            "Action":[
               "s3:GetObject"
            ],
            "Effect":"Allow",
            "Principal":{
               "AWS":[
                  "*"
               ]
            },
            "Resource":[
               "arn:aws:s3:::my-bucket/*"
            ],
            "Sid":""
         }
      ]
   }
`
```

File: docs/asset-store/docs/06-02-bucket.md [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+| **metadata.namespace** |    **NO**   | Specifies the Namespace in which the CR is available. |
+
+
+## Related resources and components
+
+These are the resources related to this CR:
+
+| Custom resource |   Description |
+|:----------:|:------|
+| Asset |  Provides the name of the storage bucket which the Asset CR refers to. |
```
Comment (pkosiec) [inline docs/asset-store/docs/06-02-bucket.md]: Hmm, is this section related to specific CR relations (references)? Because there is no relations at all. The controller can run alone, but, well, it will be much more useful along with the Asset Controller and Asset CRs.

File: docs/asset-store/docs/03-02-bucket-cr-lifecycle.md [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+![](assets/create-bucket.svg)
+
+
+## Remove a Bucket CR
+
+When you remove the Bucket CR, the Bucket Controller (BC) receives an event on the bucket deletion and removes the bucket with the whole content from Minio Gateway.
+
+The Asset Controller also monitors the status of the referenced bucket. The Asset Controller (AC) checks the Bucket CR status to make sure the bucket exists. If the bucket is deleted, the the AC receives information that that files are no longer accessible and the bucket was removed. It updates the status of the Asset CR to `ready: False` and the asset storage reference is removed. The Asset CR is not removed and you can use it later for a new bucket.
+
+![](assets/delete-bucket.svg)
```
Comment (pkosiec) [inline docs/asset-store/docs/03-02-bucket-cr-lifecycle.md]: I think we can make it more understandable . We can think about it later on Monday.

File: docs/asset-store/docs/03-02-bucket-cr-lifecycle.md [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+
+Read about the lifecycle of the Bucket Custom Resource (CR) and how other resources react to its creation and removal.
+
+
+## Create a Bucket CR
+
+When you create a Bucket CR, the Bucket Controller receives and event on the new CR and creates a bucket with the **name** specified in the CR is created in the Minio Gateway storage under the `ns-{NAMESPACE_NAME}-{CR_NAME}` location. The status of the CR contains a reference URL to the created bucket.
+
+
+![](assets/create-bucket.svg)
```
Comment (pkosiec) [inline docs/asset-store/docs/03-02-bucket-cr-lifecycle.md]: I think we can make it more understandable . We can think about it later on Monday.

File: docs/asset-store/docs/01-01-asset-store.md [hunk 11]
```diff
@@ -0,0 +1,12 @@
...
+
+The benefits of this solution are as follows:
+- It allows for storing various kinds of assets, such as markdown documents, zip, PNG, or JavaScript (JS) files.
+- It is scalable and can be used on a production system as it allows for storing assets on a large-scale using cloud provider storage services.
+- It is multi-cloud and not locked into one vendor, like AWS S3, or on-premise Minio. You can seamlessly switch between different service providers depending on your current subscription.
```
[the remark this answers is not in the document]
Comment (kazydek) [reply docs/asset-store/docs/01-01-asset-store.md]: I will think of sth ;)

File: docs/asset-store/docs/02-01-asset-store.md (hunk 5 above)
Comment (kazydek) [reply docs/asset-store/docs/02-01-asset-store.md]: As discussed, I will shorten this part and make a reference to the lifecycle documents.

File: docs/asset-store/docs/02-01-asset-store.md [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+
+![](assets/asset-store-architecture.svg)
+
+1. Create a bucket through a Bucket custom resource (CR).
+2. Create an Asset CR in which you specify the reference to the asset source location and the name of the bucket in which you want to store it.
+3. The Asset Controller (AC) receives the event on the Asset CR creation.
+4. The AC reads the CR definition and checks if the bucket is available.
+5. If it is available, the Asset Controller fetches the asset from the source location provided in the CR. If the asset is a zip or tar file, the AC unpacks the asset before uploading it onto the bucket.
+6. Optionally, the Asset Controller validates or modifies the asset if such a requirement is defined in the Asset CR. The AC communicates with the validation and mutation services and validates or modifies the asset according to the specification defined in the Asset CR.
+7. The AC uploads the asset to Minio Gateway to the bucket specified in the Asset CR.
```
[the remark this answers is not in the document]
Comment (kazydek) [reply docs/asset-store/docs/02-01-asset-store.md]: As discussed, I should leave the Minio Gateway production scenario, is that right?

File: docs/asset-store/docs/03-01-asset-cr-lifecycle.md [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+
+![](assets/create-asset.svg)
+
+## Remove an Asset CR
+
+When you remove the Asset CR, the Asset Controller receives and event on its deletion and deletes it from the Minio Gateway.
+
+![](assets/delete-asset.svg)
+
+## Modify an Asset CR
```
[the remark this answers is not in the document]
Comment (kazydek) [reply docs/asset-store/docs/03-01-asset-cr-lifecycle.md]: I will consult that with @michal-hudy

File: docs/asset-store/docs/06-01-asset.md [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+```
+apiVersion: assetstore.kyma-project.io/v1alpha1
+kind: Asset
+metadata:
+  name: my-package-assets
+  namespace: default
+spec:
+  source:
+    mode: single / index / package
```
[the remark this answers is not in the document]
Comment (kazydek) [reply docs/asset-store/docs/06-01-asset.md]: ok, I only left single

File: docs/asset-store/docs/06-01-asset.md [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+```
+
+## Custom resource parameters
+
+This table lists all the possible parameters of a given resource together with their descriptions:
+
+
+| Parameter   |      Mandatory      |  Description |
+|:----------:|:-------------:|:------|
+| **metadata.name** |    **YES**   | Specifies the name of the CR. |
```
[the remark this answers is not in the document]
Comment (kazydek) [reply docs/asset-store/docs/06-01-asset.md]: yup I know these are basics but that's one of the parameters :(

File: docs/asset-store/docs/06-02-bucket.md (hunk 3 above)
Comment (pkosiec) [reply docs/asset-store/docs/06-02-bucket.md]: This is sample of the `Bucket` CR:

```
apiVersion: assetstore.kyma-project.io/v1alpha1
kind: Bucket
metadata:
  labels:
    controller-tools.k8s.io: "1.0"
  name: test-sample
  namespace: default
spec:
  region: "us-east-1"
  policy: >
    {
      "Version":"2012-10-17",
      "Statement":[
      {
        "Effect":"Allow",
        "Principal":{
          "AWS":[
            "*"
          ]
        },
        "Action":[
          "s3:GetBucketLocation",
          "s3:ListBucket"
        ],
        "Resource":[
          "arn:aws:s3:::ns-default-test-sample"
        ]
      },
      {
        "Effect":"Allow",
        "Principal":{
          "AWS":[
            "*"
          ]
        },
        "Action":[
          "s3:GetObject"
        ],
        "Resource":[
          "arn:aws:s3:::ns-default-test-sample/*"
        ]
      }
      ]
    }
```

File: docs/asset-store/docs/06-02-bucket.md (hunk 8 above)
Comment (kazydek) [reply docs/asset-store/docs/06-02-bucket.md]: It's just a general overview of what custom resources it is / can be used in, so let's leave it here, what do you say?

File: docs/asset-store/docs/06-02-bucket.md [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+```
+kubectl get crd buckets.assetstore.kyma-project.io -o yaml
+```
+
+## Sample custom resource
+
+This is a sample resource that defines the storage bucket configuration.
+
+```
+apiVersion: assetstore.kyma-project.io/v1alpha1
```
Comment (pkosiec) [inline docs/asset-store/docs/06-02-bucket.md]: Sorry, I've given you a bad sample CR 😞 
Here's the correct one:

```yaml
apiVersion: assetstore.kyma-project.io/v1alpha1
kind: Bucket
metadata:
  labels:
    controller-tools.k8s.io: "1.0"
  name: test-sample
  namespace: default
spec:
  region: "us-east-1"
  policy: >
    {
      "Version":"2012-10-17",
      "Statement":[
      {
        "Effect":"Allow",
        "Principal":{
          "AWS":[
            "*"
          ]
        },
        "Action":[
          "s3:GetBucketLocation",
          "s3:ListBucket"
        ],
        "Resource":[
          "arn:aws:s3:::ns-default-test-sample"
        ]
      },
      {
        "Effect":"Allow",
        "Principal":{
          "AWS":[
            "*"
          ]
        },
        "Action":[
          "s3:GetObject"
        ],
        "Resource":[
          "arn:aws:s3:::ns-default-test-sample/*"
        ]
      }
      ]
    }
```

File: docs/asset-store/docs/06-02-bucket.md [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+## Custom resource parameters
+
+This table lists all the possible parameters of a given resource together with their descriptions:
+
+
+| Parameter   |      Mandatory      |  Description |
+|:----------:|:-------------:|:------|
+| **metadata.name** |    **YES**   | Specifies the name of the CR. |
+| **metadata.namespace** |    **YES**   | Specifies the Namespace in which the CR is available. |
+| **spec.location** |    **NO**   | xxx  Specifies the [location](https://docs.minio.io/docs/golang-client-api-reference.html#MakeBucket) under which the bucket is created. If the field is empty, the bucket is created under the default location. |
```
Comment (pkosiec) [inline docs/asset-store/docs/06-02-bucket.md]: Please replace the link with https://github.com/kyma-project/kyma/blob/master/components/assetstore-controller-manager/config/crds/assetstore_v1alpha1_bucket.yaml#L34

File: docs/asset-store/docs/06-02-bucket.md (hunk 8 above)
Comment (pkosiec) [reply docs/asset-store/docs/06-02-bucket.md]: Ok then 🙂

File: docs/asset-store/docs/06-01-asset.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (derberg) [reply docs/asset-store/docs/06-01-asset.md]: ok, I think that sometimes we use the name of the resource for important things, like in Bucket CR, so people might sometimes want to know what is the name used for. I agree it must stay

File: docs/asset-store/docs/06-01-asset.md [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+| **spec.source.validationwebhookservice.name** |    **NO**   | Provides the name of the validation webhook service. |
+| **spec.source.validationwebhookservice.namespace** |    **NO**   | Provides the Namespace in which the service is available. It must be the same as the asset's Namespace. |
+| **spec.source.validationwebhookservice.endpoint** |    **NO**   | Specifies the endpoint to which the service sends calls. |
+| **spec.source.mutationwebhookservice** |    **NO**   | Provides specification of the mutation webhook service. |
+| **spec.source.mutationwebhookservice.name** |    **NO**   | Provides the name of the mutation webhook service. |
+| **spec.source.mutationwebhookservice.namespace** |    **NO**   | Provides the Namespace in which the service is available. It must be the same as the asset's Namespace. |
+| **spec.source.mutationwebhookservice.endpoint** |    **NO**   | Specifies the endpoint to which the service sends calls. |
+| **spec.source.mutationwebhookservice.metadata** |    **NO**   | Provides detailed specification of the mutation service. |
```
Comment (derberg) [inline docs/asset-store/docs/06-01-asset.md]:
```suggestion
| **spec.source.mutationwebhookservice.metadata** |    **NO**   | Provides detailed metadata specific for a given mutation service and its functionality. |
```

File: docs/asset-store/docs/06-01-asset.md (hunk 2 above)
Comment (derberg) [reply docs/asset-store/docs/06-01-asset.md]: still not all status fields are described, like `assetRef`, `assetRef.assets` or `assetRef.baseUrl`

File: docs/asset-store/docs/06-02-bucket.md [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+```
+
+## Custom resource parameters
+
+This table lists all the possible parameters of a given resource together with their descriptions:
+
+
+| Parameter   |      Mandatory      |  Description |
+|:----------:|:-------------:|:------|
+| **metadata.name** |    **YES**   | Specifies the name of the CR. |
```
Comment (derberg) [inline docs/asset-store/docs/06-02-bucket.md]: not just the name of CR, this name is used as a bucket name in the storage, very important info

File: docs/asset-store/docs/06-02-bucket.md [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+| **metadata.namespace** |    **YES**   | Specifies the Namespace in which the CR is available. |
+| **spec.location** |    **NO**   | xxx  Specifies the [location](https://docs.minio.io/docs/golang-client-api-reference.html#MakeBucket) under which the bucket is created. If the field is empty, the bucket is created under the default location. |
+| **spec.policy** |    **NO**   | There is no policy specified for the bucket by default. To change it, specify the policy in a string that is similar to `{"Version": "2012-10-17","Statement": [{"Action": ["s3:GetObject"],"Effect": "Allow","Principal": {"AWS": ["*"]},"Resource": ["arn:aws:s3:::my-bucketname/*"],"Sid": ""}]}`. |
+
```
Comment (derberg) [inline docs/asset-store/docs/06-02-bucket.md]: there are also bucket statuses that need to be described here

Comment (derberg) [reply docs/asset-store/docs/06-02-bucket.md]:
```
status:
  lastHeartbeatTime: "2019-02-04T11:50:26Z"
  message: Bucket policy has been updated successfully
  phase: Ready
  reason: BucketPolicyUpdated
  url: https://minio.kyma.local/ns-default-test-sample
```

File: docs/asset-store/docs/06-01-asset.md (hunk 2 above)
Comment (kazydek) [reply docs/asset-store/docs/06-01-asset.md]: ok, I've described them

File: docs/asset-store/docs/03-03-minio-gateway.md [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+Minio is an open-source asset storage server with Amazon S3 compatible API. It can be used for storing various types of assets, such as documents, images, or videos. The size of such an asset can range from a few KBs to a maximum of 5TB.
+
+In the context of the Asset Store, the Asset Controller stores all assets in Minio, in a dedicated storage space.
+
+![](assets/minio.svg)
+
+
+## Production storage
+
+Minio does not scale for the production use due to its limits in file sizes and HTTP requests and responses. The Gateway mode, Minio's scalable counterpart, gives you more flexibility as it allows you to use the asset storage services from such major cloud providers as Azure, Amazon, and Google. Similarly to Minio, the Gateway mode
```
Comment (derberg) [inline docs/asset-store/docs/03-03-minio-gateway.md]: we cannot phrase it like this `Minio does not scale for the production use due to its limits in file sizes and HTTP requests and responses`. Minio is scalable, you can have good cluster configuration for it, there are also projects like rook.io that provide highly scalable minio configuration with dedicated operator.

we propose minio gateway for production use because:
- we aim for multi cloud solution that gives user flexibility on choosing a given cloud provider for specific kyma installation
- in future, asset store is about to support data replication and cdn configuration, and this is only possible with cloud providers
- we simply do not want to and are not interested in providing and maintaining proper minio setup for productive use as a direct storage provider

regarding `Similarly to Minio, the Gateway mode...` it is implementation detail not needed for the user as anyway he doesn't use minio directly but through the Asset CR.

File: docs/asset-store/docs/03-03-minio-gateway.md [hunk 22]
```diff
@@ -0,0 +1,31 @@
...
+
+Depending on the usage scenario, you can:
+- Use Minio for local development.
+- Store your assets on a production scale using Minio in a [Gateway mode](https://github.com/minio/minio/tree/master/docs/gateway).
+
+The Asset Store ensures that both usage scenarios work for Kyma, without additional configuration of the in-build controllers.
+
+## Local storage
+
+Minio is an open-source asset storage server with Amazon S3 compatible API. It can be used for storing various types of assets, such as documents, images, or videos. The size of such an asset can range from a few KBs to a maximum of 5TB.
```
[the remark this answers is not in the document]
Comment (kazydek) [reply docs/asset-store/docs/03-03-minio-gateway.md]: Got it and removed that part. With all due respect to Minio and its storage abilities ;)

File: docs/asset-store/docs/03-03-minio-gateway.md (hunk 21 above)
Comment (kazydek) [reply docs/asset-store/docs/03-03-minio-gateway.md]: ok, fixed as well :)

File: docs/asset-store/docs/03-03-minio-gateway.md [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+
+The whole concept of the Asset Store relies on Minio as a storage solution. It supports the Kyma's manifesto and the "batteries included" rule as Kyma provides you with this on-premise solution by default.
+
+Depending on the usage scenario, you can:
+- Use Minio for local development.
+- Store your assets on a production scale using Minio in a [Gateway mode](https://github.com/minio/minio/tree/master/docs/gateway).
+
+The Asset Store ensures that both usage scenarios work for Kyma, without additional configuration of the in-build controllers.
+
+## Local storage
```
Comment (derberg) [inline docs/asset-store/docs/03-03-minio-gateway.md]: maybe naming it `development mode storage`

File: docs/asset-store/docs/03-03-minio-gateway.md [hunk 24]
```diff
@@ -0,0 +1,31 @@
...
+
+![](assets/minio.svg)
+
+
+## Production storage
+
+For the production purposes, the Asset Store uses Minio Gateway which:
+
+- Is a multi-cloud solution that offers the flexibility to choose a given cloud provider for the specific Kyma installation, including Azure, Amazon, and Google
+- Supports data replication and CDN configuration
```
Comment (derberg) [inline docs/asset-store/docs/03-03-minio-gateway.md]: small misunderstanding
Minio Gateway doesn supports data replication and CDN configuration, the cloud providers that we can use through the gateway support that

File: docs/asset-store/docs/03-03-minio-gateway.md (hunk 23 above)
Comment (kazydek) [reply docs/asset-store/docs/03-03-minio-gateway.md]: Sure, what not ;)

File: docs/asset-store/docs/03-03-minio-gateway.md (hunk 24 above)
Comment (kazydek) [reply docs/asset-store/docs/03-03-minio-gateway.md]: ok, clear, thanks

File: docs/asset-store/docs/03-02-bucket-cr-lifecycle.md [hunk 25]
```diff
@@ -0,0 +1,23 @@
...
+title: Bucket custom resource lifecycle
+type: Details
+---
+
+Learn about the lifecycle of the Bucket custom resource (CR) and how its creation and removal affects other Asset Store components.
+
+
+## Create a Bucket CR
+
+When you create a Bucket CR, the Bucket Controller (BC) receives an Event on the new CR and creates a bucket with the name specified in the CR. It is created in the Minio Gateway storage under the `ns-{NAMESPACE_NAME}-{CR_NAME}` location. The status of the CR contains a reference URL to the created bucket.
```
Comment (bszwarc) [inline docs/asset-store/docs/03-02-bucket-cr-lifecycle.md]:
```suggestion
When you create a Bucket CR, the Bucket Controller (BC) receives a CR creation Event and creates a bucket with the name specified in the CR. It is created in the Minio Gateway storage under the `ns-{NAMESPACE_NAME}-{CR_NAME}` location. The status of the CR contains a reference URL to the created bucket.
```

File: docs/asset-store/docs/03-02-bucket-cr-lifecycle.md [hunk 26]
```diff
@@ -0,0 +1,23 @@
...
+
+
+![](assets/create-bucket.svg)
+
+
+## Remove a Bucket CR
+
+When you remove the Bucket CR, the BC receives an Event on the bucket deletion and removes the bucket with the whole content from Minio Gateway.
```
Comment (bszwarc) [inline docs/asset-store/docs/03-02-bucket-cr-lifecycle.md]:
```suggestion
When you remove the Bucket CR, the BC receives CR deletion Event and removes the bucket with the whole content from Minio Gateway.
```

File: docs/asset-store/docs/03-02-bucket-cr-lifecycle.md [hunk 27]
```diff
@@ -0,0 +1,23 @@
...
+
+The Asset Controller (AC) also monitors the status of the referenced bucket. The AC checks the Bucket CR status to make sure the bucket exists. If the bucket is deleted, the AC receives information that that files are no longer accessible and the bucket was removed. The AC updates the status of the Asset CR to `ready: False` and the asset storage reference is removed. The Asset CR is not removed and you can use it later for a new bucket.
```
Comment (bszwarc) [inline docs/asset-store/docs/03-02-bucket-cr-lifecycle.md]:
```suggestion
The Asset Controller (AC) also monitors the status of the referenced bucket. The AC checks the Bucket CR status to make sure the bucket exists. If the bucket is deleted, the AC receives information that the files are no longer accessible and the bucket was removed. The AC updates the status of the Asset CR to `ready: False` and the asset storage reference is removed. The Asset CR is not removed and you can use it later for a new bucket.
```

File: docs/asset-store/docs/03-03-minio-gateway.md [hunk 28]
```diff
@@ -0,0 +1,32 @@
+---
+title: Minio and Minio Gateway
+type: Details
+---
+
+The whole concept of the Asset Store relies on Minio as a storage solution. It supports the Kyma's manifesto and the "batteries included" rule as Kyma provides you with this on-premise solution by default.
```
Comment (bszwarc) [inline docs/asset-store/docs/03-03-minio-gateway.md]:
```suggestion
The whole concept of the Asset Store relies on Minio as the storage solution. It supports the Kyma's manifesto and the "batteries included" rule as Kyma provides you with this on-premise solution by default.
```

File: docs/asset-store/docs/03-03-minio-gateway.md [hunk 29]
```diff
@@ -0,0 +1,32 @@
...
+
+Depending on the usage scenario, you can:
+- Use Minio for local development.
+- Store your assets on a production scale using Minio in a [Gateway mode](https://github.com/minio/minio/tree/master/docs/gateway).
+
+The Asset Store ensures that both usage scenarios work for Kyma, without additional configuration of the in-build controllers.
+
+## Development mode storage
+
+Minio is an open-source asset storage server with Amazon S3 compatible API. It can be used for storing various types of assets, such as documents, files, or images.
```
Comment (bszwarc) [inline docs/asset-store/docs/03-03-minio-gateway.md]:
```suggestion
Minio is an open-source asset storage server with Amazon S3 compatible API. You can use it to store various types of assets, such as documents, files, or images.
```

File: docs/asset-store/docs/03-03-minio-gateway.md [hunk 30]
```diff
@@ -0,0 +1,32 @@
...
+
+![](assets/minio.svg)
+
+
+## Production storage
+
+For the production purposes, the Asset Store uses Minio Gateway which:
+
+- Is a multi-cloud solution that offers the flexibility to choose a given cloud provider for the specific Kyma installation, including Azure, Amazon, and Google
+- Allows for the use of various cloud providers that support the data replication and CDN configuration
```
Comment (bszwarc) [inline docs/asset-store/docs/03-03-minio-gateway.md]:
```suggestion
- Allows you to use various cloud providers that support the data replication and CDN configuration
```

File: docs/asset-store/docs/06-01-asset.md [hunk 31]
```diff
@@ -0,0 +1,121 @@
...
+  bucketRef:
+    name: my-bucket
+
+```
+
+### Validation and mutation webhook services
+
+You can also define validation and mutation services:
+- **Validation webhook** performs the validation of fetched assets before they are uploaded to the bucket. It can be a list of several different validation webhooks and all of them should be processed even if one fails. It can refer either to the validation of a specific file against a specification or to the security validation. The validation webhook returns the validation status.
+- **Mutation webhook** acts similarly to the validation service. The difference is that it mutates the asset instead of just validating it. For example, this can be an asset rewriting through the `regex` operation or `keyvalue`, or the modification in the JSON specification. The mutation webhook returns modified files instead of information on the status.
```
Comment (bszwarc) [inline docs/asset-store/docs/06-01-asset.md]:
```suggestion
- **Mutation webhook** acts similarly to the validation service. The difference is that it mutates the asset instead of just validating it. For example, this can mean asset rewriting through the `regex` operation or `keyvalue`, or the modification in the JSON specification. The mutation webhook returns modified files instead of information on the status.
```

File: docs/asset-store/docs/03-03-minio-gateway.md [hunk 32]
```diff
@@ -0,0 +1,32 @@
...
+- Is also compatible with Amazon S3 APIs
```
Comment (bszwarc) [inline docs/asset-store/docs/03-03-minio-gateway.md]:
```suggestion
- Is compatible with Amazon S3 APIs
```

File: docs/asset-store/docs/03-02-bucket-cr-lifecycle.md [hunk 33]
```diff
@@ -0,0 +1,23 @@
+---
+title: Bucket custom resource lifecycle
+type: Details
+---
+
+Learn about the lifecycle of the Bucket custom resource (CR) and how its creation and removal affects other Asset Store components.
```
Comment (bszwarc) [inline docs/asset-store/docs/03-02-bucket-cr-lifecycle.md]:
```suggestion
Learn about the lifecycle of the Bucket custom resource (CR) and how its creation and removal affect other Asset Store components.
```

File: docs/asset-store/docs/02-01-asset-store.md [hunk 34]
```diff
@@ -0,0 +1,37 @@
...
+
+- **Mutation Service** is an optional service which ensures that the asset is modified according to the mutation specification defined in the Asset CR before it is uploaded to the bucket. The service returns the modified asset to the AC.
+
+- **Minio Gateway** is a Minio cluster mode which is a production-scalable storage solution. It ensures flexibility of using asset storage services from major cloud providers, including Azure Blob Storage, Amazon S3, and Google Cloud Storage.
+
+## Asset flow
+
+This diagram provides an overview of the basic Asset Store workflow and the role of particular components in this process:
+
+![](assets/asset-store-architecture.svg)
```
[the remark this answers is not in the document]
Comment (kazydek) [reply docs/asset-store/docs/02-01-asset-store.md]: true

File: docs/asset-store/docs/03-01-asset-cr-lifecycle.md [hunk 35]
```diff
@@ -0,0 +1,30 @@
...
+
+## Remove an Asset CR
+
+When you remove the Asset CR, the AC receives an Event on its deletion and deletes it from Minio Gateway.
+
+![](assets/delete-asset.svg)
+
+## Change the bucket reference
+
+When you modify an Asset CR by updating the bucket reference in the Asset CR to a new one, while the previous bucket still exists, the lifecycle starts again. The asset is created in a new storage location and its new location is updated in the Asset CR.
```
[the remark this answers is not in the document]
Comment (kazydek) [reply docs/asset-store/docs/03-01-asset-cr-lifecycle.md]: As discussed, I will leave `starts again`.

File: docs/asset-store/docs/03-01-asset-cr-lifecycle.md [hunk 36]
```diff
@@ -0,0 +1,30 @@
...
+
+## Create an Asset CR
+
+When you create an Asset CR, the Asset Controller (AC) receives a CR creation Event, reads its definition, verifies if the bucket exists, downloads the asset, unpacks it, and stores it in Minio Gateway.
+
+![](assets/create-asset.svg)
+
+## Remove an Asset CR
+
+When you remove the Asset CR, the AC receives a CR deletion Event and deletes it from Minio Gateway.
```
Comment (bszwarc) [inline docs/asset-store/docs/03-01-asset-cr-lifecycle.md]:
```suggestion
When you remove the Asset CR, the AC receives a CR deletion Event and deletes the CR from Minio Gateway.
```

File: docs/asset-store/docs/03-01-asset-cr-lifecycle.md [hunk 37]
```diff
@@ -0,0 +1,30 @@
+---
+title: Asset custom resource lifecycle
+type: Details
+---
+
+Learn about the lifecycle of the Asset custom resource (CR) and how its creation, removal, or a change in the bucket reference affects other Asset Store components.
+
+## Create an Asset CR
+
+When you create an Asset CR, the Asset Controller (AC) receives a CR creation Event, reads its definition, verifies if the bucket exists, downloads the asset, unpacks it, and stores it in Minio Gateway.
```
Comment (bszwarc) [inline docs/asset-store/docs/03-01-asset-cr-lifecycle.md]:
```suggestion
When you create an Asset CR, the Asset Controller (AC) receives a CR creation Event, reads the CR definition, verifies if the bucket exists, downloads the asset, unpacks it, and stores it in Minio Gateway.
```
